client/src/controllers/InitController.ts

```diff
@@ -41,15 +62,30 @@ export class InitController extends Controller<HTMLElement> {
    * Support the ability to also dispatch custom event names.
    */
   ready() {
-    const events = this.eventValue.split(' ').filter(Boolean);
     const delayValue = this.delayValue;
+    const detail = { ...(this.detailValue || {}) }; // Coerce to object
+    const events = this.eventValue.split(' ').filter(Boolean);
 
     debounce(() => true, delayValue < 0 ? null : delayValue)().then(() => {
       this.element.classList.add(...this.readyClasses);
       this.element.classList.remove(...this.removeClasses);
-      this.dispatch('ready', { bubbles: true, cancelable: false });
+
+      const readyEvent = this.dispatch('ready', {
+        bubbles: true,
+        cancelable: false,
+        detail,
+      });
+      if (readyEvent?.defaultPrevented) {
+        return; // return early if preventDefault() is called
+      }
+
       events.forEach((name) => {
-        this.dispatch(name, { bubbles: true, cancelable: false, prefix: '' });
+        this.dispatch(name, {
+          bubbles: true,
+          cancelable: false,
+          prefix: '',
+          detail,
```

Comment:
```suggestion
          cancelable: false,
          detail,
          prefix: '',
```

Minor nitpick, can we make prefix last so it's alpha sorted.